File: backend/app/visualizations/community_insights.py
import random
from datetime import datetime, timedelta

class CommunityInsightsVisualizer:
    """Creates an interactive visualization of community insights."""

    def __init__(self):
        self.resource_categories = ['Documentation', 'Tutorials', 'Blog Posts', 'Videos', 'Workshops']
        self.engagement_types = ['Views', 'Comments', 'Shares', 'Likes']
        self._generate_sample_data()

    def _generate_sample_data(self):
        """Generate sample data for resource impact analysis."""
        self.data = []
        
        # Generate 20 data points
        for _ in range(20):
            category = random.choice(self.resource_categories)
            engagement = random.choice(self.engagement_types)
            
            data_point = {
                'category': category,
                'engagement_type': engagement,
                'effectiveness_score': random.uniform(25, 70),  # Score between 25-70%
                'reach': random.randint(100, 1000),
                'timestamp': (datetime.now() - timedelta(days=random.randint(0, 30))).isoformat()
            }
            self.data.append(data_point)

    def generate_visualization_data(self):
        """Generate visualization data for the frontend."""
        return {
            'scatter_data': self.data,
            'categories': self.resource_categories,
            'engagement_types': self.engagement_types,
            'summary': {
                'total_resources': len(self.data),
                'avg_effectiveness': sum(d['effectiveness_score'] for d in self.data) / len(self.data),
                'total_reach': sum(d['reach'] for d in self.data),
                'time_range': {
                    'start': min(d['timestamp'] for d in self.data),
                    'end': max(d['timestamp'] for d in self.data)
                }
            }
        }

File: frontend/src/lib/visualization-generators/community-insights.ts
const RESOURCE_CATEGORIES = ['Documentation', 'Tutorials', 'Blog Posts', 'Videos', 'Workshops']
const ENGAGEMENT_TYPES = ['Views', 'Comments', 'Shares', 'Likes']

function randomChoice<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)]
}

function randomUniform(min: number, max: number): number {
  return min + Math.random() * (max - min)
}

function randomInt(min: number, max: number): number {
  return Math.floor(min + Math.random() * (max - min))
}

export function generateCommunityInsights() {
  const data = Array.from({ length: 20 }, () => {
    const daysAgo = randomInt(0, 30)
    const date = new Date()
    date.setDate(date.getDate() - daysAgo)

    return {
      category: randomChoice(RESOURCE_CATEGORIES),
      engagement_type: randomChoice(ENGAGEMENT_TYPES),
      effectiveness_score: randomUniform(25, 70),
      reach: randomInt(100, 1000),
      timestamp: date.toISOString(),
    }
  })

  return {
    scatter_data: data,
    categories: RESOURCE_CATEGORIES,
    engagement_types: ENGAGEMENT_TYPES,
    summary: {
      total_resources: data.length,
      avg_effectiveness: data.reduce((sum, d) => sum + d.effectiveness_score, 0) / data.length,
      total_reach: data.reduce((sum, d) => sum + d.reach, 0),
      time_range: {
        start: data.reduce((min, d) => (d.timestamp < min ? d.timestamp : min), data[0].timestamp),
        end: data.reduce((max, d) => (d.timestamp > max ? d.timestamp : max), data[0].timestamp),
      },
    },
  }
}
